### DISCUSSION TITLE
Given JSON object, get its schema definition

### DISCUSSION THREAD

**Discussion by rsinohara** [colinhacks/zod#585] [Q&A]
Hi, I am just starting with zod, looks amazing.

The one tool I am missing and would dearly need is something that can generate a schema from a given json object. Doesn't have to be extra smart (but some intelligence would be great).

Is there any such tool around? I will eventually be using this so much that it might save time if I have to do a rudimentary one.

To be clear, I don't need that dynamically, just static conversion to get started on the schema.

**Comment by scotttrinh** [comment]
Are you picturing a command-line tool that outputs a Zod schema given some arbitrary JSON input?

```
$ cat input.json
{
  "a": "a string",
  "b": 42,
  "c": true,
  "d": null,
  "xs": ["x", "y", "z"]
}
$ cat input.json | some-zod-cli-tool
z.object({
  a: z.string(),
  b: z.number(),
  c: z.boolean(),
  d: z.null(),
  xs: z.array(z.string()),
});
```

Something like that? I'm not aware of anything like that, but there are projects output JSON schemas so perhaps you could use something like https://github.com/StefanTerdell/zod-to-json-schema as inspiration for making such a tool?

**Comment by StefanTerdell** [comment]
Here you go :)

```typescript
import { format } from "prettier";

const jsonToZod = (obj: any): string => {
  const parse = (obj: any): string => {
    switch (typeof obj) {
      case "string":
        return "z.string()";
      case "number":
        return "z.number()";
      case "bigint":
        return "z.number().int()";
      case "boolean":
        return "z.boolean()";
      case "object":
        if (Array.isArray(obj)) {
          const options = obj
            .map(parse)
            .reduce(
              (acc, curr) => (acc.includes(curr) ? acc : [...acc, curr]),
              []
            );
          if (options.length === 1) {
            return `z.array(${options[0]})`;
          } else if (options.length > 1) {
            return `z.array(z.union([${options}]))`;
          } else {
            return "z.array(z.unknown())";
          }
        }

        return `z.object({${Object.entries(obj)
          .map(([k, v]) => `${k}:${parse(v)}`)
          }})`;
      case "undefined":
        return "z.undefined()";
      case "function":
        return "z.function()";
      case "symbol":
      default:
        return "z.unknown()";
    }
  };

  return format(`const schema=${parse(obj)}`, { parser: "babel" });
};
```

**Reply by rsinohara** [reply]
Wow amazing. Is this available somewhere? Seems like it could be running to make it easier to use. If not, you mind I put it up?

**Reply by StefanTerdell** [reply]
Sure, have at it. Wouldn't mind [a contributor tag](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#people-fields-author-contributors) if you do make something of it.

**Reply by StefanTerdell** [reply]
Actually, I got bored and made a package

https://www.npmjs.com/package/json-to-zod

**Reply by rsinohara** [reply]
You get full credit, unless I add something substantial 👍

**Reply by rsinohara** [reply]
Ah, I wanted to create a little page and wrap it, would be a bit easier for my use case (I need to do many small to medium objects).

**Reply by StefanTerdell** [reply]
Clone and run locally :)

https://github.com/StefanTerdell/json-to-zod-react

**Reply by StefanTerdell** [reply]
Probably too lazy to maintain it though. Let me know if I should transfer the repos/npm package rights to you and you can go from there.

**Reply by StefanTerdell** [reply]
Turns out Github Pages is insanely easy to work with :D 

https://stefanterdell.github.io/json-to-zod-react/

**Reply by rsinohara** [reply]
Yes, it's incredibly easy, that's what I was going for :)

**Reply by rsinohara** [reply]
Amazing!!

**Comment by StefanTerdell** [comment]
Reposting my answer so it doesn't get buried in the thread

[Here's a page where you can paste your JSON or JS objects (JSON5)](https://rsinohara.github.io/json-to-zod-react/) 

[Or if you prefer a package/CLI](https://www.npmjs.com/package/json-to-zod)

If anyone wants to maintain it, let me know and I'll transfer the repos.

[edit: borken link]

**Reply by rsinohara** [reply]
Amazing! This should probably be linked. I'll be using it for my project, so I'll contribute if I see any thing that can be added. But honestly, it's pretty useful!!
Thanks!

**Reply by rsinohara** [reply]
FYI I am typing _many_ objects in my application.

**Reply by rsinohara** [reply]
I PR'ed adding it to the zod's readme. I can maintain it if you prefer, I will do some fixes as needed.

**Reply by StefanTerdell** [reply]
Great! I'll look into how to transfer ownership. What's your username on NPM?

**Reply by rsinohara** [reply]
https://www.npmjs.com/settings/renatosinohara/profile

**Reply by rsinohara** [reply]
You can just give me access if you prefer, instead of transferring. Maybe that is what you meant.

**Reply by colinhacks** [reply]
Thanks @StefanTerdell! Looks great.

**Reply by StefanTerdell** [reply]
@rsinohara Nah, I'll transfer them completely. I have no use for them so it's better you own them. Just one thing, you have to delete your fork before I can transfer the package repo to you. You should already have gotten the website repo, though.

**Reply by rsinohara** [reply]
@StefanTerdell fork deleted. I got the mail for the page repo, but it expired before I could use the link.

**Reply by StefanTerdell** [reply]
@rsinohara Ok, I'll try and send both again now

**Reply by rsinohara** [reply]
Got them. I just need to put github page up again. Thanks for the work! I'll add you, if not already there.

**Reply by StefanTerdell** [reply]
Cool! Looks online to me https://rsinohara.github.io/json-to-zod-react/

You might want to PR the URL change to the Zod repo.

`npm run deploy` to update it.

GLHF! :)
